### PR TITLE
Removes unnecessary database calls

### PR DESCRIPTION
**Pull request by zegnus**
### Problem description

The method `updateLocked` is fired really fast several times in a row, inside we have a loop through all the batches and for every batch there is an extra database call in order to find out if the batch has started for the first time.
### Solution implemented

All the information that we need is loaded from the database every time at the beginning of the `updateLocked` method, a second call to the same database is unnecessary as all the information is loaded in memory.

We have replaced the database call for just looping through the loaded items.

**Logs**

```
11-18 17:48:16.533 7981-9664/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:346).updateLocked] updateLocked isBatchStartingForTheFirstTime: true 
11-18 17:48:16.553 7981-9664/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:16.562 7981-9664/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.987 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:346).updateLocked] updateLocked isBatchStartingForTheFirstTime: false 
11-18 17:48:56.987 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 165574089 
11-18 17:48:56.988 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.988 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.989 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.989 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.989 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.990 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.991 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.992 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.993 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.994 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.995 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.997 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.998 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.998 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.998 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.998 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.999 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.999 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.999 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:56.999 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.000 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.000 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.000 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.000 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.001 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.001 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.001 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.001 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.001 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.002 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.002 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.002 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0 
11-18 17:48:57.003 7981-10269/? V/NoTils: [DownloadManager-UpdateThread][(DownloadService.java:352).updateLocked] updateLocked fileDownloadInfo.getCurrentBytes(): 0
```

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java [hunk 1]
```diff
@@ -375,8 +374,17 @@ private boolean updateLocked() {
         return isActive;
     }
 
+    private boolean isBatchStartingForTheFirstTime(DownloadBatch downloadBatch) {
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: will this still be consistent data? or can it get out of sync?

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: Still consistent as the check is done on top of the amount of bytes downloaded, equivalent to the old database check

File: library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java [hunk 2]
```diff
@@ -24,8 +24,7 @@
 
     private static final String VERSION_ONE_TO_VERSION_TWO_MIGRATION_SCRIPT = "ALTER TABLE "
             + DownloadContract.Batches.BATCHES_TABLE_NAME
-            + " ADD "
-            + DownloadContract.Batches.COLUMN_HAS_STARTED + " BOOLEAN NOT NULL DEFAULT 0;";
+            + ";";
```
Comment (takecare) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: so, do we still need this migration? not much happening here

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: that is true, we don't, good point

Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: :-1:  can't just do this I'm afraid, you'll need to create another migration that removes the column

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java (hunk 1 above)
Comment (takecare) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: given it no longer goes to the db i imagine you can get not the most up to date data but should not be a big deal (if it actually runs this way). am i wrong?

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: Corner case covered here https://github.com/novoda/download-manager/commit/0ff7a1d0c2a0a04244eca5039b0606b603603168

File: library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java (hunk 2 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: Migration covered here https://github.com/novoda/download-manager/commit/92c7b0823ccaa309448fa4b6f50622b05cb22538 there is no alter with removing columns, we have to drop and recreate

File: library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java [hunk 3]
```diff
@@ -113,18 +111,13 @@ public void onCreate(@NonNull SQLiteDatabase db) {
...
 
-    private void upgradeFromVersionOneToVersionTwo(@NonNull SQLiteDatabase db) {
-        try {
-            db.execSQL(VERSION_ONE_TO_VERSION_TWO_MIGRATION_SCRIPT);
-        } catch (SQLException ex) {
-            LLog.e("couldn't update table in downloads database to v2");
-            throw ex;
-        }
+    private void upgradeFromVersionTwoToVersionThree(@NonNull SQLiteDatabase db) {
+        createBatchesTable(db);
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: shouldn't this do an `alter table` like it used to do? calling the expert @ouchadam

Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: calling this method will drop the table :cry:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java [hunk 4]
```diff
@@ -341,8 +345,8 @@ private boolean updateLocked() {
             }
 
             if (!isActive && downloadReadyChecker.canDownload(downloadBatch)) {
-                boolean isBatchStartingForTheFirstTime = batchRepository.isBatchStartingForTheFirstTime(downloadBatch.getBatchId());
-                if (isBatchStartingForTheFirstTime) {
+                if (isBatchStartingForTheFirstTime(downloadBatch)) {
+                    Log.v("Ferran, bach started the first time");
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: :x:

File: library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java]: I thought that I put a comment on this, there is no alter for removing a column, there are three options:
1. Drop
2. Create a temporary table, copy values, drop tmp table
3. Leave this column and is not used

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java [hunk 5]
```diff
@@ -375,8 +378,30 @@ private boolean updateLocked() {
         return isActive;
     }
 
+    private boolean isBatchStartingForTheFirstTime(DownloadBatch downloadBatch) {
+        boolean isBatchWithoutWrittenBytes = downloadBatch.getCurrentSize() == 0;
+
+        long batchId = downloadBatch.getBatchId();
+        if (batchWithoutWrittenBytesAlreadyChecked(isBatchWithoutWrittenBytes, batchId)) {
```
Comment (Mecharyry) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: Maybe I am missing something but I'm not sure why you need this, could you explain this for me?

To me, it looks like you should return in the statements below.

**Comment by takecare** [comment]
what's up with this pr?

**Comment by blundell** [comment]
closing till someone who cares re-opens

**Pull request closed without merging**